### PR TITLE
bpo-42972: Fully implement GC protocol for ssl heap types (GH-26370)

### PR DESCRIPTION
**Pull request by erlend-aasland**
https://bugs.python.org/issue42972

### CODE REVIEW

File: Modules/_ssl.c [hunk 1]
```diff
@@ -434,11 +434,11 @@ static PyType_Slot sslerror_type_slots[] = {
...
-    "ssl.SSLError",
-    sizeof(PyOSErrorObject),
-    0,
-    Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE | Py_TPFLAGS_IMMUTABLETYPE,
-    sslerror_type_slots
+    .name = "ssl.SSLError",
+    .basicsize = sizeof(PyOSErrorObject),
+    .flags = (Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE |
+              Py_TPFLAGS_IMMUTABLETYPE | Py_TPFLAGS_HAVE_GC),
+    .slots = sslerror_type_slots
```
Comment (erlend-aasland) [inline Modules/_ssl.c:441]: Let me know if you want me to revert the designated initialiser changes. Ditto for all types.

Comment (tiran) [reply Modules/_ssl.c:441]: I like them, good work :)

Comment (erlend-aasland) [reply Modules/_ssl.c:441]: Phew :) Thanks!

**Review by tiran — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks, @tiran !

**Comment by pablogsal** [comment]
This PR introduced a regression: it added a GC type without fully implementing the GC protocol. See:

https://bugs.python.org/issue44252

**Comment by pablogsal** [comment]
This PR *also* has introduce reference leaks in the urllib module:

dcb8786a9848516e823e090bb36079678913d8d3 is the first bad commit
commit dcb8786a9848516e823e090bb36079678913d8d3
Author: Erlend Egeberg Aasland <erlend.aasland@innova.no>
Date:   Thu May 27 09:50:09 2021 +0200

    bpo-42972: Fully implement GC protocol for ssl heap types (GH-26370)

 Modules/_ssl.c | 73 ++++++++++++++++++++++++++++++++++++----------------------
 1 file changed, 45 insertions(+), 28 deletions(-)

example failure:

https://buildbot.python.org/all/#/builders/320/builds/41

**Comment by pablogsal** [comment]
Fixed by https://bugs.python.org/issue44252

**Pull request merged**